README.rst
===============================
Swagger Fuzzer
===============================

Swagger Fuzzer helps you do fuzzing testing on your Swagger APIs.

* Free software: MIT License
* Documentation: https://swagger_fuzzer.readthedocs.org.

Usage
-----

Install swagger-fuzzer via pip:

.. code:: shell

    pip install https://github.com/Lothiraldan/swagger-fuzzer/archive/master.zip

Point it to your swagger specification:

.. code:: shell

    swagger-fuzzer http://localhost:80/v2/swagger.json

If swagger-fuzzer find any problem, it will print you a report, for example:

.. code:: shell

    Falsifying example: swagger_fuzzer(data=request(...))
    Curl command: curl -i -X GET -d '' 'http://localhost/pet/findByStatus'
    Traceback (most recent call last):
      File "/Users/lothiraldan/.virtualenvs/swagger-fuzzer/bin/swagger-fuzzer", line 9, in <module>
        load_entry_point('swagger-fuzzer', 'console_scripts', 'swagger-fuzzer')()
      File "/Users/lothiraldan/projects/swagger-fuzzer/swagger_fuzzer/swagger_fuzzer/swagger_fuzzer.py", line 34, in main
        do(args)
      File "/Users/lothiraldan/projects/swagger-fuzzer/swagger_fuzzer/swagger_fuzzer/swagger_fuzzer.py", line 79, in do
        swagger_fuzzer()
      File "/Users/lothiraldan/projects/swagger-fuzzer/swagger_fuzzer/swagger_fuzzer/swagger_fuzzer.py", line 68, in swagger_fuzzer
        @hsettings(max_examples=settings.iterations)
      File "/Users/lothiraldan/projects/swagger-fuzzer/hypothesis-python/src/hypothesis/core.py", line 541, in wrapped_test
        print_example=True, is_final=True
      File "/Users/lothiraldan/projects/swagger-fuzzer/hypothesis-python/src/hypothesis/executors.py", line 58, in default_new_style_executor
        return function(data)
      File "/Users/lothiraldan/projects/swagger-fuzzer/hypothesis-python/src/hypothesis/core.py", line 104, in run
        return test(*args, **kwargs)
      File "/Users/lothiraldan/projects/swagger-fuzzer/swagger_fuzzer/swagger_fuzzer/swagger_fuzzer.py", line 76, in swagger_fuzzer
        validator(SPEC, request, result, settings)
      File "/Users/lothiraldan/projects/swagger-fuzzer/swagger_fuzzer/swagger_fuzzer/validators.py", line 54, in valid_output_mime
        raise AssertionError("Response content-type {} is not declared: {}".format(response.headers['Content-Type'], valids))
    AssertionError: Response content-type text/html; charset=ISO-8859-1 is not declared: []

Bugs hunted
-----------

Swagger-fuzzer really find bugs, here is a non-extensive list of bugs found with swagger-fuzzer:

* https://github.com/pallets/flask/issues/1761, null byte filename crash flask file handling code.

Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

This project was sponsored by my employer Sqreen_ and developed initially for our needs.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _Sqreen: https://www.sqreen.io
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2016-05-20)
------------------

* First release on PyPI.


tests/test_swagger_fuzzer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_swagger_fuzzer
----------------------------------

Tests for `swagger_fuzzer` module.
"""


import unittest

from swagger_fuzzer import swagger_fuzzer


class TestSwagger_fuzzer(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_000_something(self):
        pass


if __name__ == '__main__':
    import sys
    sys.exit(unittest.main())


swagger_fuzzer/strategy.py
""" Hypothesis custom strategy
"""
from hypothesis.searchstrategy import SearchStrategy
from hypothesis.control import _current_build_context
from hypothesis.errors import InvalidArgument
from hypothesis.control import note


def data():
    """This isn't really a normal strategy, but instead gives you an object
    which can be used to draw data interactively from other strategies.

    It can only be used within @given, not find. This is because the lifetime
    of the object cannot outlast the test body.

    See the rest of the documentation for more complete information.

    """
    class DataObject(object):

        def __init__(self, data):
            self.count = 0
            self.data = data
            context = _current_build_context.value
            context.request = {}

        def __repr__(self):
            return 'request(...)'

        def draw(self, strategy, name=None):
            result = self.data.draw(strategy)
            return result

    class DataStrategy(SearchStrategy):
        supports_find = False

        def do_draw(self, data):
            if not hasattr(data, 'hypothesis_shared_data_strategy'):
                data.hypothesis_shared_data_strategy = DataObject(data)
            return data.hypothesis_shared_data_strategy

        def __repr__(self):
            return 'data()'

        def map(self, f):
            self.__not_a_first_class_strategy('map')

        def filter(self, f):
            self.__not_a_first_class_strategy('filter')

        def flatmap(self, f):
            self.__not_a_first_class_strategy('flatmap')

        def example(self):
            self.__not_a_first_class_strategy('example')

        def __not_a_first_class_strategy(self, name):
            raise InvalidArgument((
                'Cannot call %s on a DataStrategy. You should probably be '
                "using @composite for whatever it is you're trying to do."
            ) % (name,))
    return DataStrategy()


swagger_fuzzer/swagger_fuzzer.py
# -*- coding: utf-8 -*-
""" Swagger Fuzzer helps you do fuzzing testing on your Swagger APIs.
"""
import argparse
from urllib.parse import urlparse, urlunparse

import requests
from hypothesis import given, settings as hsettings, note
from swagger_spec_validator.util import get_validator

from .strategy import data
from .validators import VALIDATORS
from .swagger_helpers import get_request

parser = argparse.ArgumentParser()
parser.add_argument('spec_url', help="The Swagger spec url")
parser.add_argument('-n', '--number', dest='iterations',
                    default=1000000, type=int,
                    help='Maximum number of iterations (default: 100000)')
parser.add_argument('-s', '--standard-http-code', dest='http_code',
                    action='append', type=int,
                    help='Standards http codes, no need to declare them in '
                         'swagger spec, default: 200, 404, 405')


def main():
    args = parser.parse_args()
    if args.http_code is None:
        args.http_code = [200, 405, 404]
    do(args)


def to_curl_command(request):
    """ Convert a requests preparred request to curl command
    """
    command = "curl -i -X {method}{headers} -d '{data}' '{uri}'"
    method = request.method
    uri = request.url
    data = request.body
    if data is None:
        data = ''
    headers = ["{0}: '{1}'".format(k, v) for k, v in request.headers.items()]
    headers = " -H ".join(headers)
    if headers:
        headers = " -H {} ".format(headers)
    return command.format(method=method, headers=headers, data=data, uri=uri)


def do(settings):
    PARSED_HOST = urlparse(settings.spec_url)

    swagger_spec = requests.get(settings.spec_url)
    swagger_spec.raise_for_status()
    SPEC = swagger_spec.json()

    validator = get_validator(SPEC, settings.spec_url)
    validator.validate_spec(swagger_spec.json(), settings.spec_url)

    SPEC_HOST = urlunparse(list(PARSED_HOST)[:2] + [SPEC['basePath']] + ['', '', ''])

    s = requests.Session()

    @given(data())
    @hsettings(max_examples=settings.iterations)
    def swagger_fuzzer(data):
        request = get_request(data, SPEC, SPEC_HOST)
        note("Curl command: {}".format(to_curl_command(request)))

        result = s.send(request)

        for validator in VALIDATORS:
            validator(SPEC, request, result, settings)

    # Call the function
    swagger_fuzzer()

if __name__ == '__main__':
    main()
    # swagger_fuzzer()


swagger_fuzzer/swagger_helpers.py
""" Helpers for swagger specifications
"""
import json
import hypothesis.strategies as st
from hypothesis.extra.datetime import datetimes
from requests import Request
from furl import furl
from hypothesis import assume

from .utils import CustomJsonEncoder

SWAGGER_FORMAT_MAPPING = {
    'int64': st.integers(),
    'string': st.text(),
    'integer': st.integers(),
    'int32': st.integers(),
    'date-time': datetimes(),
    'boolean': st.booleans(),
}


def _is_swagger_parameter(dict):
    """ Check if the parameter dict is a valid swagger parameter
    """
    return dict.get('type') or dict.get('schema') or dict.get('$ref')


def get_ref(ref, spec):
    assert ref.startswith('#/')
    splitted = ref.lstrip('#/').split('/')

    referenced_to = spec

    for path in splitted:
        referenced_to = referenced_to[path]

    return referenced_to


def get_item_path_acceptable_format(path_item, spec):
    if path_item.get('consumes'):
        return path_item['consumes']

    return spec.get('consumes')


def _get_filtered_parameter(path_item, in_, spec):
    parameters = path_item.get('parameters')
    filtered_params = [p for p in parameters if p['in'] == in_]
    non_converted_params = {p['name']: p for p in filtered_params}
    return CustomTransformation(get_ref, spec).transform(non_converted_params)


def get_request(data, spec, spec_host):
    endpoint_path = data.draw(st.sampled_from(spec['paths'].keys()))
    endpoint = spec['paths'][endpoint_path]

    method_name = data.draw(st.sampled_from(endpoint.keys()))
    endpoint = endpoint[method_name]

    path_params = _get_filtered_parameter(endpoint, 'path', spec)
    path_args = data.draw(st.fixed_dictionaries(path_params))

    query_params = _get_filtered_parameter(endpoint, 'query', spec)
    query_args = data.draw(st.fixed_dictionaries(query_params))

    body_params = _get_filtered_parameter(endpoint, 'body', spec)
    if body_params:
        body_args = data.draw(body_params['body'])
    else:
        body_args = None

    valid_request_body_format = get_item_path_acceptable_format(endpoint, spec)

    request_data = None
    request_headers = {}

    if body_args:
        # no_body_format_declaration(body_args, valid_request_body_format, endpoint)
        if body_args and valid_request_body_format is None:
            # Force a request format, swagger ui seems to force json format
            valid_request_body_format = ["application/json"]

        request_body_format = data.draw(st.sampled_from(valid_request_body_format), 'request_body_format')

        request_headers['Content-Type'] = request_body_format
        if request_body_format == 'application/x-www-form-urlencoded':
            request_data = body_args
        elif request_body_format == 'application/json':
            request_data = json.dumps(body_args, cls=CustomJsonEncoder)
        elif request_body_format == 'application/xml':
            pass
            # TODO Implement XML
        else:
            raise Exception(request_body_format)

    endpoint_url = endpoint_path.format(**path_args)
    assume('\x00' not in endpoint_url)

    # Generate request
    URL = furl(spec_host)
    URL = URL.join(endpoint_url.lstrip('/'))

    if query_args:
        URL = URL.add(args=query_args)

    request = Request(method_name, URL.url, data=request_data,
                      headers=request_headers).prepare()
    request.build_context = locals()
    return request


class CustomTransformation(object):

    def __init__(self, get_ref, spec):
        self.get_ref = get_ref
        self.spec = spec

    def transform(self, obj):
        """
        """
        obj = self._recursive_transform(obj)
        if isinstance(obj, (list, tuple)):
            return self._transform_array(obj)
        if isinstance(obj, dict):
            return self._transform_dict(obj)
        else:
            return self._transform_obj(obj)

    def _transform_array(self, obj):
        new_array = []
        for index, value in enumerate(obj):
            new_array.append(self.transform(value))

        return new_array

    def _transform_dict(self, obj):
        new_dict = {}
        for key, value in obj.items():
            new_dict[key] = self.transform(value)

        return new_dict

    def _transform_obj(self, obj):
        return self.default(obj)

    def _recursive_transform(self, obj):
        old_obj = None

        while old_obj is not obj:
            old_obj = obj
            obj = self.default(obj)

        return obj

    def default(self, obj):
        """
        """
        if isinstance(obj, dict) and _is_swagger_parameter(obj):
            parameter_type = obj.get('format', obj.get('type'))
            parameter_schema = obj.get('schema')
            parameter_ref = obj.get('$ref')

            if parameter_type in SWAGGER_FORMAT_MAPPING:
                return SWAGGER_FORMAT_MAPPING[parameter_type]
            elif parameter_ref:
                return self.transform(self.get_ref(parameter_ref, self.spec))
            elif parameter_type == 'array':
                if obj['items'].get('enum'):
                    return st.lists(elements=st.sampled_from(obj['items']['enum']))
                elif obj['items'].get('type'):
                    return st.lists(elements=SWAGGER_FORMAT_MAPPING[obj['items']['type']])
                elif obj['items'].get('$ref'):
                    schema = self.get_ref(obj['items']['$ref'], self.spec)
                    return st.lists(elements=self.transform(schema))
                raise Exception('array', obj)
            elif parameter_type == 'object':
                properties = {}
                for property_name, property_ in obj['properties'].items():
                    properties[property_name] = self.transform(property_)
                return st.fixed_dictionaries(properties)
            elif parameter_schema:
                if parameter_schema.get('type') == 'array':
                    schema = self.get_ref(parameter_schema['items']['$ref'], self.spec)
                    return st.lists(elements=self.transform(schema))
                else:
                    schema = self.get_ref(parameter_schema['$ref'], self.spec)
                    transformed = self.transform(schema)
                    return transformed
            else:
                raise Exception("Invalid", obj, parameter_type)

        return obj


swagger_fuzzer/validators.py
""" Validators
"""


def check_result_status_code(spec, request, response, settings):
    """ Check that response status code is either a "standard" one
    like 404, 405, 200 (use -s cli argument to change it) or one
    of the declared one globally or for the path in swagger configuration
    """
    status_code = int(response.status_code)
    endpoint_path = request.build_context['endpoint_path']
    authorized = spec['paths'][endpoint_path][request.method.lower()]['responses'].keys()

    # Default means all status code are allowed
    if "default" in authorized:
        return

    allowed = set(settings.http_code).union(map(int, authorized))

    if status_code not in allowed:
        raise AssertionError("Request on {!r} returned status_code {}, not in declared one {}".format(request.url, response.status_code, list(allowed)))


def no_server_error(spec, request, response, settings):
    """ Check that response status code is different than 500
    """
    if response.status_code == 500:
        raise AssertionError("Request on {!r} returns status_code {}".format(URL, response.status_code))


def no_body_format_declaration(spec, request, response, settings):
    """ Check that for each post path, a body format is declared
    """
    body_args = request.build_context.get('body_args')
    if request.build_context['body_args'] and request.build_context.get('request_body_format') is None:
        raise AssertionError("Body parameters but no declared format for endpoint {}: {}".format(endpoint, body_args))


def valid_output_mime(spec, request, response, settings):
    """ Check that each request returns with a content-type that is declared
    """
    global_valids = spec.get('consumes', [])

    endpoint_path = request.build_context['endpoint_path']
    path = spec['paths'][endpoint_path][request.method.lower()]
    local_valids = path.get('consumes', [])

    if local_valids:
        valids = local_valids
    else:
        valids = global_valids

    if response.headers['Content-Type'] not in valids:
        raise AssertionError("Response content-type {} is not declared: {}".format(response.headers['Content-Type'], valids))


VALIDATORS = [
    no_server_error,
    no_body_format_declaration,
    check_result_status_code,
    valid_output_mime
]
